Hello Michael!

My name is Olga. I'm happy to review your project today. 

When I will see mistake at the first time, I will just point it out. I let you find it and fix it by yourself. I'm trying to prepare you to work as an Data Analyst. SO, at a real job, your team lead will do the same. But if you can't handle this task yet, I will give you a more accurate hint at the next iteration.

You will find my comments below - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Success. Everything is done succesfully.
</div>


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Remarks. Some recommendations.
</div>


<div class="alert alert-block alert-danger">
    
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
    
Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>


You can answer me by using this: 

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment</b> <a class="tocSkip"></a>
    
Michael, thank you for sending your project. You've done a really good job on it!

While there's room for improvement, on the whole, your project is impressive good.

<span class="text-danger">There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix. Could you check my comments? Do not hesistate to ask your tutor for help, if you have any problems!  </span>
    
Every issue with our code is a chance for us to learn something new.

Good luck! :)
    
</div>

<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Hello Michael! Thank you for your corrections! All queries are correct now.
    
I'm glad to say that your project has been accepted.
    
Keep up the good work, and good luck with the last project!</div>

# SQL Project
by MICHAEL BIRMAN

## Task
- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating: look only at books with at least 50 ratings.
- Find the average number of text reviews among users who rated more than 50 books.

In [1]:
!pip install psycopg2-binary

In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

                                       Let's look at the data

In [3]:
query = """
    SELECT table_name, column_name, data_type
    FROM information_schema.columns
    WHERE table_schema = 'public'
    ORDER BY table_name, ordinal_position;
"""
df = pd.read_sql(query, con=engine)

# print the DataFrame
df

table_name       column_name                    data_type
0   advertisment_costs                id                      integer
1   advertisment_costs          sourceid                     smallint
2   advertisment_costs                dt  timestamp without time zone
3   advertisment_costs             costs                        money
4              authors         author_id                      integer
5              authors            author                         text
6                books           book_id                      integer
7                books         author_id                      integer
8                books             title                         text
9                books         num_pages                      integer
10               books  publication_date                         date
11               books      publisher_id                      integer
12              orders                id                      integer
13              orders             buyts  timestamp without time zone
14              orders           revenue                        money
15              orders               uid            character varying
16          publishers      publisher_id                      integer
17          publishers         publisher                         text
18             ratings         rating_id                      integer
19             ratings           book_id                      integer
20             ratings          username                         text
21             ratings            rating                      integer
22             reviews         review_id                      integer
23             reviews           book_id                      integer
24             reviews          username                         text
25             reviews              text                         text
26              visits                id                      integer
27              visits               uid            character varying
28              visits            device            character varying
29              visits             endts  timestamp without time zone
30              visits          sourceid                     smallint
31              visits           startts  timestamp without time zone

We see the names of the tables, the fields in them and the type of fields

            Let's look at the tables themselves and display the first five records

In [4]:
tables = ['advertisment_costs', 'authors', 'books', 'orders', 'publishers', 'ratings', 'reviews', 'visits']

for table in tables:
    query = f"SELECT * FROM {table} LIMIT 5"
    df = pd.read_sql(query, con=engine)
    print(f"Table: {table}\n")
    display(df)

Table: advertisment_costs



id  sourceid         dt   costs
0   1         1 2017-06-01  $75.20
1   2         1 2017-06-02  $62.25
2   3         1 2017-06-03  $36.53
3   4         1 2017-06-04  $55.00
4   5         1 2017-06-05  $57.08

Table: authors



author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Table: books



book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Table: orders



id               buyts revenue                   uid
0   1 2017-06-01 00:10:00  $17.00  10329302124590727494
1   2 2017-06-01 00:25:00   $0.55  11627257723692907447
2   3 2017-06-01 00:27:00   $0.37  17903680561304213844
3   4 2017-06-01 00:29:00   $0.55  16109239769442553005
4   5 2017-06-01 07:58:00   $0.37  14200605875248379450

Table: publishers



publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Table: ratings



rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Table: reviews



review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Table: visits



id                   uid   device               endts  sourceid  \
0   0  16879256277535980062    touch 2017-12-20 17:38:00         4   
1   1    104060357244891740  desktop 2018-02-19 17:21:00         2   
2   2   7459035603376831527    touch 2017-07-01 01:54:00         5   
3   3  16174680259334210214  desktop 2018-05-20 11:23:00         9   
4   4   9969694820036681168  desktop 2017-12-27 14:06:00         3   

              startts  
0 2017-12-20 17:20:00  
1 2018-02-19 16:53:00  
2 2017-07-01 01:54:00  
3 2018-05-20 10:59:00  
4 2017-12-27 14:06:00

There is information about advertisment costs, authors, books, orders, publishers ,ratings,visits, as well as user reviews of books. The relationship between tables with data is shown in the diagram:

In [5]:
from IPython.display import Image
Image(url='https://practicum-content.s3.us-west-1.amazonaws.com/resources/moved_Untitled_-_2020-07-02T142019.920_1593688954.png')

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Great, we have an introduction and connection! Also it was informative to add scheme of our database.</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

And we should check duplicates and missing values.</div>


## Task 1
Find the number of books released after January 1, 2000.

In [6]:
query = """
SELECT COUNT(*) FROM books WHERE publication_date > '2000-01-01';
"""
df = pd.read_sql(query, con=engine)

# print the DataFrame

df


count
0    819

We have 819 books released after January 1, 2000

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Correct!</div>

## Task 2:
Find the number of user reviews and the average rating for each book.

In [7]:

query = """

SELECT 
  b.title,
  COUNT(DISTINCT rev.review_id) AS num_reviews, 
  ROUND(AVG(rat.rating)::numeric, 2) AS avg_rating
FROM 
  books AS b
  LEFT JOIN ratings AS rat USING(book_id)
  LEFT JOIN reviews AS rev USING(book_id)
GROUP BY 
  b.title
ORDER BY 
  num_reviews DESC;
"""
df = pd.read_sql(query, con=engine)

# print the DataFrame
df


title  num_reviews  \
0                                  Memoirs of a Geisha            8   
1                              Twilight (Twilight  #1)            7   
2    Harry Potter and the Chamber of Secrets (Harry...            6   
3                                     The Glass Castle            6   
4                                      Eat  Pray  Love            6   
..                                                 ...          ...   
994    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
995                            The Natural Way to Draw            0   
996   The Cat in the Hat and Other Dr. Seuss Favorites            0   
997                          Essential Tales and Poems            0   
998  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_rating  
0          4.14  
1          3.66  
2          4.29  
3          4.21  
4          3.40  
..          ...  
994        3.67  
995        3.00  
996        5.00  
997        4.00  
998        4.00  

[999 rows x 3 columns]

the book with the most reviews is Memoirs of a Geisha with 8 reviews and a 4.14 average rating. there are also books that don't have reviews but are rated. Let's look at them separately

<div class="alert alert-block alert-danger">
<s><b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Please, pay attention on calculations of num_reviews.</div>

<div class="alert alert-block alert-info">
<b>Done.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Correct!</div>

## Task 3
Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [9]:

query = """
SELECT 
  publishers.publisher,
  COUNT(books.book_id) AS num_books
FROM 
  books
  INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE 
  books.num_pages > 50
GROUP BY 
  publishers.publisher
ORDER BY 
  num_books DESC;
"""
df = pd.read_sql(query, con=engine)

# print the DataFrame

df



publisher  num_books
0                                        Penguin Books         42
1                                              Vintage         31
2                             Grand Central Publishing         25
3                                     Penguin Classics         24
4                                     Ballantine Books         19
..                                                 ...        ...
329                                         Turtleback          1
330  Atheneum Books for Young Readers: Richard Jack...          1
331                                     Penguin Signet          1
332                                    Victor Gollancz          1
333                      Harvard Business Review Press          1

[334 rows x 2 columns]

 We see that we have 334 such editions.  The leader is Penguin Books with 42 published books over 50 pages.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
You are right!</div>

## Task 4:
Identify the author with the highest average book rating: look only at books with at least 50 ratings.

In [10]:


query = """
SELECT 
  authors.author, 
  ROUND(AVG(ratings.rating), 2) AS avg_rating
FROM 
  ratings
  INNER JOIN books ON ratings.book_id = books.book_id
  INNER JOIN authors ON books.author_id = authors.author_id
WHERE 
  ratings.book_id IN (
    SELECT 
      book_id 
    FROM 
      ratings 
    GROUP BY 
      book_id 
    HAVING 
      COUNT(*) >= 50
  )
GROUP BY 
  authors.author
ORDER BY 
  avg_rating DESC
LIMIT 5;
"""
df = pd.read_sql(query, con=engine)

# print the DataFrame

df

author  avg_rating
0         J.K. Rowling/Mary GrandPré        4.29
1  Markus Zusak/Cao Xuân Việt Khương        4.26
2                     J.R.R. Tolkien        4.25
3                  Louisa May Alcott        4.19
4                       Rick Riordan        4.08


The author with the highest average rating J.K. Rowling and Mary GrandPre

<div class="alert alert-block alert-danger">
<s><b>Reviewer's comment</b> <a class="tocSkip"></a>
    
J.K. Rowling/Mary GrandPré is correct. But please check calculations of avg_rating.</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Well done!</div>

## Task 5:
Find the average number of text reviews among users who rated more than 50 books.

In [11]:
           
 
query = """
SELECT
  ROUND(AVG(Sub_rev.reviews_avg)::numeric, 2) AS users_avg_reviews
FROM
  (SELECT
    rev.username AS username,
    COUNT(rev.review_id) AS reviews_avg
  FROM
    reviews rev
  WHERE
    rev.username IN
      (SELECT
        Sub_rat.username
      FROM
        (SELECT
          rat.username AS username,
          COUNT(rat.rating_id) AS ratings_num
        FROM
          ratings rat
        GROUP BY
          rat.username
        HAVING
          COUNT(rat.rating_id) > 50) AS Sub_rat)
  GROUP BY
    rev.username) AS Sub_rev;
"""
df = pd.read_sql(query, con=engine)

# print the DataFrame

df           

users_avg_reviews
0              24.33

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Perfect! </div>

# Conclusion
In summary, our database contains information on 1000 books, with 819 of them, or about 80%, released in the last 20 years.

We created a reference table with the number of reviews and the average rating for each book, which shows the popularity and quality of the book in the eyes of the readers. It is worth noting that books with the highest number of reviews do not always have the highest average rating (for example, "Twilight" with 1120 reviews and only a 3.7 average rating).

We found that the most active publisher is Penguin Books, which has released 42 books with more than 50 pages in our dataset. And the author with the highest average rating is J.K. Rowling, the author of the Harry Potter series. It is likely that any book she releases will attract a lot of attention.

Finally, we learned that even the most active readers, who rated at least 50 books, wrote an average of about 24 textual reviews. This means that the likelihood of an active user writing a textual review of a book is quite low. For regular readers, the chances are even lower.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Thak you for the conclusion! </div>